[release-4.12] NE-1372: Add support for AWS shared VPC in another account #966

gcs278 opened the pull request:
Add support for configuring DNS records in AWS Route 53 using a separate account for the private hosted zone.

This PR cherry-picks https://github.com/openshift/cluster-ingress-operator/pull/928/commits/7251aaa9a30f3759c2731d061207d2f63cd945de (from https://github.com/openshift/cluster-ingress-operator/pull/928, the original implementation) and https://github.com/openshift/cluster-ingress-operator/pull/951/commits/047bd986587a923029586a2497ec99d0a5a242a0 (from https://github.com/openshift/cluster-ingress-operator/pull/951, a bug fix for https://issues.redhat.com/browse/OCPBUGS-14998).

**How I generated this backport:**
1. Created a new commit to vendor openshift/api backport https://github.com/openshift/api/pull/1529
    - Modified `go.mod` to use `github.com/openshift/api@v0.0.0-20230817133225-564be9ddb58e` (from https://github.com/openshift/api/pull/1551)
    - `go mod tidy`
    - `go mod vendor`
    - `make` to regenerate bindata.go
2. `git cherry-pick 7251aaa9a30f3759c2731d061207d2f63cd945de` (https://github.com/openshift/cluster-ingress-operator/pull/928/commits/7251aaa9a30f3759c2731d061207d2f63cd945de)
3. Resolved merge conflicts:
    - `pkg/manifests/bindata.go`: Regenerated
    - `pkg/operator/controller/dns/controller.go`:  Removed [`PrivateHostedZoneAWSEnabled`](https://github.com/openshift/cluster-ingress-operator/blob/b4b5f424e49a31840fd133561e01178a5d123e19/pkg/operator/controller/dns/controller.go#L118) from config structure for the dns controller. Removed associated check for [`r.config.PrivateHostedZoneAWSEnabled`](https://github.com/openshift/cluster-ingress-operator/blob/b4b5f424e49a31840fd133561e01178a5d123e19/pkg/operator/controller/dns/controller.go#L675C3-L675C61)
    - `pkg/operator/operator.go`:  https://github.com/openshift/cluster-ingress-operator/pull/928/commits/7251aaa9a30f3759c2731d061207d2f63cd945de had feature gate logic. We are not backporting the feature gate so I did not integrate any feature gate logic in the merge.
4. Completed cherry-pick with add, `git cherry-pick --continue`, etc...
5. `git cherry-pick 047bd986587a923029586a2497ec99d0a5a242a0` (https://github.com/openshift/cluster-ingress-operator/pull/951/commits/047bd986587a923029586a2497ec99d0a5a242a0)
    - No merge conflicts

melvinjoseph86 commented:
retest-required

melvinjoseph86 commented:
/retest-required

melvinjoseph86 commented:
/retest-required

patrickdillon commented on manifests/00-ingress-credentials-request.yaml:
```diff
@@ -26,6 +26,7 @@ spec:
       - route53:ListTagsForResources
       - route53:ChangeResourceRecordSets
       - tag:GetResources
+      - sts:AssumeRole
```
Let's remove this a la:
https://github.com/openshift/cluster-ingress-operator/pull/972

gcs278 replied:
yep done.

gcs278 commented:
Since Miciah has been helping with these backports:
/assign @Miciah

sdodson commented:
/retest-required

gcs278 commented:
https://github.com/openshift/cluster-ingress-operator/compare/0a774bc0ce151b1b7a4052e0399cde62470aea1d..fb01b1991085a7e55c0f45998b723c353d03fc3a

Just commit message updates:
- Added API version in https://github.com/openshift/cluster-ingress-operator/pull/971/commits/d321d8f7ceb265a2757b58fbc35aa0e0a7609865 (was `TBD`)
- Moved credentials request drop of `sts:assume-role` to https://github.com/openshift/cluster-ingress-operator/pull/971/commits/f4c49cbfb4c720138cebbeae2c5c59744a7ab74a, which added it, not https://github.com/openshift/cluster-ingress-operator/pull/971/commits/fb01b1991085a7e55c0f45998b723c353d03fc3a (where I put the removal)
- Updated the commit message of https://github.com/openshift/cluster-ingress-operator/pull/971/commits/f4c49cbfb4c720138cebbeae2c5c59744a7ab74a to reflect the modifications in this backport, something I forgot to do for 4.13. The commit message had details of the feature gate, and credentials requests updates, which I removed, since those parts were dropped.

Miciah commented:
/approve
/lgtm

This change does not change the operator's behavior unless the new `spec.platform.aws.privateZoneIAMRole` option in the cluster DNS config is specified.  Note that specifying `spec.platform.aws.privateZoneIAMRole` could cause the operator to fail to update DNS records if the operator does not have the `sts:AssumeRole` permission.  However, adding the `sts:AssumeRole` permission to the credentials request in a z-stream update can cause other problems (see [OCPBUGS-17733](https://issues.redhat.com/browse/OCPBUGS-17733) and https://github.com/openshift/cluster-ingress-operator/pull/972), so rather than modifying the credentials request, we will have a release note to warn against using the new API field without giving the operator the required permissions.  (In addition, we might amend https://github.com/openshift/api/pull/1551 and update https://github.com/openshift/cluster-config-operator/pull/344 to add a warning to the godoc for `spec.platform.aws.privateZoneIAMRole`, but even if we do amend the godoc, that change only needs to be vendored in cluster-config-operator, not necessarily cluster-ingress-operator, so that change needn't block this PR.)
/label backport-risk-assessed

@gcs278, as far as I am concerned, the hold can be removed once you have the release note sorted out.

melvinjoseph86 commented:
/label cherry-pick-approved

melvinjoseph86 commented:
/jira refresh

gcs278 commented:
/unhold
